lib/ckb/tx/transaction.rb

```diff
@@ -0,0 +1,179 @@
+# frozen_string_literal: true
+
+require 'secp256k1'
+
+require_relative '../blake2b'
+require_relative 'out_point'
+require_relative 'input'
+require_relative 'script'
+require_relative 'output'
+
+module Ckb
+  class Transaction
+    attr_reader :version, :deps, :inputs, :outputs, :witnesses
+
+    def initialize(version: 0, deps: [], inputs: [], outputs: [], witnesses: [])
+      @version = version
+      @deps = deps
+      @inputs = inputs
+      @outputs = outputs
+      @witnesses = witnesses
+    end
+
+    def self.from_h(h)
+      return h if h.is_a?(Transaction)
+
+      new(
+        version: h[:version],
+        deps: h[:deps],
+        inputs: h[:inputs].map { |i| Input.from_h(i) },
+        outputs: h[:outputs].map { |o| Output.from_h(o) },
+        witnesses: h[:witnesses] || []
+      )
+    end
+
+    def sign_sighash_all_inputs(key)
+      signed_inputs = self.class.sign_sighash_all_inputs(@inputs, @outputs, key.privkey)
+      self.class.new(
+        version: version,
+        deps: deps,
+        inputs: signed_inputs,
+        outputs: outputs,
+        witnesses: witnesses
+      )
+    end
+
+    # @return [Ckb::Input[]]
+    def self.sign_sighash_multiple_anyonecanpay_inputs(inputs, outputs, privkey)
+      sign_sighash_inputs(inputs, outputs, privkey, 0x84)
+    end
+
+    # @return [Ckb::Input[]]
+    def self.sign_sighash_all_anyonecanpay_inputs(inputs, outputs, privkey)
+      sign_sighash_inputs(inputs, outputs, privkey, 0x81)
+    end
+
+    # In Ruby, bytes are represented using String, since JSON has no native byte arrays,
+    # CKB convention bytes passed with a "0x" prefix hex encoding, hence we
+    # have to do type conversions here.
+    def normalize_for_json!
+      @inputs.each do |input|
+        input.args = input.args.map do |arg|
+          Ckb::Utils.bin_to_hex(
+            Ckb::Utils.delete_prefix(arg)
+          )
+        end
+      end
+
+      @outputs.each do |output|
+        output.data = Ckb::Utils.bin_to_hex(
+          Ckb::Utils.delete_prefix(output.data)
+        )
+        output.lock.args = output.lock.args.map do |arg|
+          Ckb::Utils.bin_to_hex(
+            Ckb::Utils.delete_prefix(arg)
+          )
+        end
+
+        next unless output.type
+
+        output.type.args = output.type.args.map do |arg|
+          Ckb::Utils.bin_to_hex(
```

Comment:
Why don't we do the `delete_prefix` inside `bin_to_hex`? Does `bin_to_hex` accepts `hex` string both with our with `0x` and produce different result? (If so please ignore this question)

Reply:
I don't see any `0x` prefixed string in `bin_to_hex`, just a little strange `delete_prefix` inside `bin_to_hex`